### ISSUE TITLE
Nested cross-validation example in docs seems to be not properly nesting CV as per scikit-learn documentation

### ISSUE THREAD

**Issue by bjsmith** [nilearn/nilearn#2883]
scikit-learn guidelines describe the typical motivation for nested cross-validation. They [say](https://scikit-learn.org/stable/auto_examples/model_selection/plot_nested_cross_validation_iris.html) that

> Model selection without nested CV uses the same data to tune model parameters and evaluate model performance. Information may thus “leak” into the model and overfit the data. 

### Actual documentation

The problematic nilearn cv example is here http://nilearn.github.io/auto_examples/02_decoding/plot_haxby_grid_search.html in section 9.3.10.4. It describes nested cross-validation with this example:

```
from sklearn.model_selection import KFold
cv = KFold(n_splits=3)
nested_cv_scores = []

for train, test in cv.split(session):
    y_train = np.array(y)[train]
    y_test = np.array(y)[test]
    val_scores = []

    for sp in screening_percentile_range:
        decoder = Decoder(estimator='svc', mask=mask_img,
                          smoothing_fwhm=4, cv=3, standardize=True,
                          screening_percentile=sp)
        decoder.fit(index_img(fmri_niimgs, train), y_train)
        y_pred = decoder.predict(index_img(fmri_niimgs, test))
        val_scores.append(np.mean(y_pred == y_test))

    nested_cv_scores.append(np.max(val_scores))

print("Nested CV score: %.4f" % np.mean(nested_cv_scores))
```

Model hyper-parameter tuning here, i.e., selecting an optimal `screening_percentile_range`, is based on the performance of the hyper-parameter in predicting the test set data (the 'selection' takes place at the point where the maximum item in `val_scores` is selected using `np.max`). But this same test set data is used to evaluate the final performance. Thus, it does exactly what the scikit-learn docmentation says non-nested CV does, i.e., "uses the same data to tune model parameters and evaluate model performance".

This is not a trivial problem because it could lead users to believe they have properly implemented nested-CV and avoided overfitting, when fact they have implemented non-nested CV and retain overfitting.

### Expected documenation

For proper nested CV, I think the range of `screening_percentile_range` values should be passed in to the `Decoder` object's `param_grid` parameter.

Then, selection of `screening_percentile_range` would be truly independent from the final validation score at the end. Modified, the example above might look like:
```
from sklearn.model_selection import KFold
cv = KFold(n_splits=3)
nested_cv_scores = []

for train, test in cv.split(session):
    y_train = np.array(y)[train]
    y_test = np.array(y)[test]
    val_scores = []

    decoder = Decoder(estimator='svc', mask=mask_img,
                      smoothing_fwhm=4, cv=3, standardize=True,
                      param_grid={'screening_percentile'=screening_percentile_range})
    decoder.fit(index_img(fmri_niimgs, train), y_train)
    y_pred = decoder.predict(index_img(fmri_niimgs, test))
    val_scores.append(np.mean(y_pred == y_test))

    nested_cv_scores.append(np.max(val_scores))

print("Nested CV score: %.4f" % np.mean(nested_cv_scores))
```

This actually wouldn't entirely address use cases in instances where someone wants to, e.g., compare different estimators using nested cross-validation, because it's not possible to pass different estimators into the `param_grid` parameter. And as a guide to exactly what's going on it's a bit insufficient (the 'nesting' is hidden away within the Decoder function's cross-validation). But it would be technically correct for a more limited use-case.

### Further comments

Further, I'd request a feature to allow integration of estimator selection within the inner CV loop built into the decoder, or alternatively, permit the Decoder to run internally without cross-validation so that this can be arranged externally.

There's a chance I'm mistaken on this report and I hope someone else could verify what I'm saying is correct, but I am fairly confident the nilearn documentation is inconsistent with the scikit-learn guidelines.

**Comment by jeromedockes** [comment]
indeed taking the max of scores biases the measured accuracy. @tbng do you want to take a look?

regarding the further comments, the point of the Decoder is that it uses the estimators fitted during hyperparameter selection to build an ensemble model. it averages their coefficients, which can only be done if they correspond to the same estimator. 
about running without cross-validation, you could force it not to do cross-validation by passing an appropriate `cv` cross-validation object, but that would disable its main functionality so it may be just as simple to chain the univariate feature selection and linear model in a scikit-learn Pipeline

**Comment by bjsmith** [comment]
Yes I considered forcing it not to do cross-validation. The only option that seemed viable was to use `model_selection.PredefinedSplit(test_fold)`. This would force it to use a consistent cross-validation. One could then run the `DecoderRegressor.fit` multiple times for each estimator or other set of parameters one wanted to consider, and the splits would at least be consistent. But that seems inefficient because of the number of times I'm rerunning `DecoderRegressor.fit`.

**Comment by jeromedockes** [comment]
"Ben J. Smith" ***@***.***> writes:

> Yes I considered forcing it not to do cross-validation. The only option that seemed viable was to use `model_selection.PredefinedSplit(test_fold)`.

you could also pass an iterator that yields a single train, test pair of
indices, with the whole data in train, or define a class that provides
the n_splits and get_n_splits methods -- see the scikit-learn documentation.

> This would force it to use a consistent cross-validation. One could then run the `DecoderRegressor.fit` multiple times for each estimator or other set of parameters one wanted to consider, and the splits would at least be consistent. But that seems inefficient because of the number of times I'm rerunning `DecoderRegressor.fit`.

the estimator will be fitted for each set of parameters anyway, whether
you do it manually or the Decoder does it. but it is true that by
fitting the whole Decoder each time you may do the univariate selection more
times than necessary (if you are using it)

in any case if you do not want to benefit from the Decoder's ensembling
of models fitted in each fold, and all you need is a grid search for the
best hyperparameters, the simplest solution is to not use the Decoder,
and use scikit-learn's GridSearchCV and Pipeline.

**Comment by thomasbazeille** [comment]
@bjsmith Just a broad statement that is maybe unclear from the documentation : the Decoder object is not designed to simplify the run of a nice cross-validation and return you the best overall the best set of paramaters while refitting a model using those. 

It is not either a proper way to compare several estimators and draw statistical conclusions on the one you should be using. 

It is rather an efficient bagging implementation that will try many things on each fold and just average all the best models together, often yielding stable and strong accuracy.

Another solution that may work to compare parameter impacts while avoiding recoding it would be to fit one Decoder for each parameter set..

**Comment by tbng** [comment]
> indeed taking the max of scores biases the measured accuracy. @tbng do you want to take a look?
> 
> regarding the further comments, the point of the Decoder is that it uses the estimators fitted during hyperparameter selection to build an ensemble model. it averages their coefficients, which can only be done if they correspond to the same estimator.
> about running without cross-validation, you could force it not to do cross-validation by passing an appropriate `cv` cross-validation object, but that would disable its main functionality so it may be just as simple to chain the univariate feature selection and linear model in a scikit-learn Pipeline

Thanks @bjsmith a lot for your feedback. 

It seems indeed doing the `nested_cv_scores.append(np.max(val_scores))` does not make sense for the example as in this case the Decoder would overfit the test data. Do you want to open PR for improvement of the example?

For feature request and use case where one would not want to use cross-validation, I'm not sure it then the `Decoder` would remains to its original purpose of creating simple `do-it-all` pipeline for end-user that does not care about technical details in between, as @thomasbazeille  said.

Lastly IMO adding the model/estimator selection right inside the `Decoder` would come at the cost of much more computational time -- along with implementing other hyperparameter tuning, etc; so I'm really not sure about it.

**Comment by bthirion** [comment]
The Decoder object should be able to take a pipeline and run nested cross-val on the Pipeline's params ? Currently this is not the case as far as I can see.

**Comment by jeromedockes** [comment]
> The Decoder object should be able to take a pipeline and run nested cross-val on the Pipeline's params ?

I don't think so because the Decoder averages the coefficients of a
linear model fitted several times, on several subsets of the training
data.
So it couldn't accept an arbitrary pipeline.

of course it would be easy to store all fitted estimators and average
their predictions rather than coefficients, but that is not what the
current implementation does (and possibly requires more memory,
depending on the estimators that are used)

**Comment by bthirion** [comment]
Point taken. 
But then we should reintroduce somewhere in the examples a case in which we perform nested cross-val for parameter selection in a pipeline. This is an important use case I think.

**Comment by jeromedockes** [comment]
> But then we should reintroduce somewhere in the examples a case in which we perform nested cross-val for parameter selection in a pipeline. This is an important use case I think.

that's true. there is this one:

  https://nilearn.github.io/auto_examples/07_advanced/plot_age_group_prediction_cross_val.html

that uses a GridSearch and a Pipeline, although it is for age prediction
with connectivity rather than for decoding.

**Comment by bthirion** [comment]
Indeed ! We should better reference it from the decoding section. 
An the example could be more pedagogic by explaining the strategy used (note the nested GridSerchCV call).

**Comment by tbng** [comment]
The Decoder also already can taken a grid (technically a dictionary) of hyperparameter for an estimator (either supported or custom one) and do a CV for selecting the best one, IMO this is very close to a sklearn pipeline with GridSearchCV. 

Again I'm not sure adding feature to run a estimators selection inside the Decoder is a good idea.

**Comment by bthirion** [comment]
Simply remember that users often consider anova+svm as a good solution (which it is indeed), and that this unavoidably requires to tune the number of features selected by Anova ---hence the need of nested CV. It is actually disturbing that doing this through a Decoder is not possible. We need to find a way to explain why.

**Comment by bjsmith** [comment]
I hear the points that the Decoder is not designed to do principled tuning of different estimators. Having said that, it seems to be designed to test different parameters of each particular estimator, and conceptually, from an end user's point of view, perhaps there's not much difference? But I wouldn't advocate overhauling the logic or the decoder if this doesn't really fit technically.

 @tbng I will get started with that PR but when I have done it, hopefully someone can review my contribution?

**Comment by bjsmith** [comment]
> Another solution that may work to compare parameter impacts while avoiding recoding it would be to fit one Decoder for each parameter set..

I considered writing a CV algorithm that wraps the decoder, fitting one decoder for each parameter set. but that seems quite inefficient because within the decoder itself one can't avoid running CV. So it seemed inefficient to do a full CV loop around the decoder which itself does CV.

Having said that, perhaps the CV the decoder is doing should be considered part of the training process rather than optimizing hyper-parameters, in which case, it always makes sense to use CV within the decoder even when you are running a separate hyper-parameter optimization outside the Decoder. Is that right?

**Comment by bjsmith** [comment]
@tbng do you think the suggested replacement in the OP makes sense for the documentation? I thought so when I posted it, but since hearing from others on here that the inner CV loop within the Decoder isn't designed to do parameter selection, I'm unsure. Having said that I can't see why you couldn't put the parameter selection in the Decoder's CV loop along with the other bagging that it does.

**Comment by man-shu** [comment]
Here's the newer example that breaks down what the Decoder does: https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_understand_decoder.html

Would this help close this issue?

**Comment by michellewang** [comment]
I took a deeper look at the current version of the [example](https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_grid_search.html#compute-prediction-scores-with-different-values-of-screening-percentile) in the documentation.

It does look like the hyperparameter tuning is not done correctly, as the code block below does not do anything to pick the best value for the `screening_percentile` hyperparameter, but rather takes the average of the best scores across folds:

```python
import warnings

from sklearn.exceptions import ConvergenceWarning
from sklearn.model_selection import KFold

cv = KFold(n_splits=3)
nested_cv_scores = []

for train, test in cv.split(run):
    y_train = np.array(y)[train]
    y_test = np.array(y)[test]
    val_scores = []

    for sp in screening_percentile_range:
        with warnings.catch_warnings():
            # silence warnings about Liblinear not converging
            # increase the number of iterations.
            warnings.filterwarnings(
                action="ignore", category=ConvergenceWarning
            )
            decoder = Decoder(
                estimator="svc",
                mask=mask_img,
                smoothing_fwhm=4,
                cv=3,
                standardize="zscore_sample",
                screening_percentile=sp,
                param_grid=param_grid,
                verbose=1,
            )
        decoder.fit(index_img(fmri_niimgs, train), y_train)
        y_pred = decoder.predict(index_img(fmri_niimgs, test))
        val_scores.append(np.mean(y_pred == y_test))

    nested_cv_scores.append(np.max(val_scores))

print(f"Nested CV score: {np.mean(nested_cv_scores):.4f}")
```

What we want instead is something like this:
- Outer CV loop for model evaluation
- Inner CV loop for hyperparameter tuning, similar idea as `sklearn`'s `GridSearchCV` but we can't use that directly since it's not compatible with the `Decoder`s (due to taking brain images as input)
- Note: in addition to the above, there is the additional CV being done by the decoder itself, which is separate. That is being done only on the classifier part of the decoder, and instead of doing typical CV hyperparameter tuning, the CV is used to create an ensemble model (the final classifier is an average of the best classifiers for each CV fold).

In general I am tempted to revamp the entire example to try to improve clarity, and to better differentiate between the different CV loops and the built-in CV in the decoder.

**Comment by michellewang** [comment]
Regarding https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_understand_decoder.html:
1. It doesn't look like the tuning of feature selection hyperparameters (e.g. `screening_percentile`) is addressed, so I think we still need that nested CV example
2. The ensemble model part is not really mentioned, maybe we should add some text at the end explaining the final `Decoder` model and which parameters are used in `Decoder.predict`
